crates/core/src/surfnet/mod.rs

```diff
@@ -0,0 +1,696 @@
+use crate::{
+    rpc::utils::convert_transaction_metadata_from_canonical,
+    types::{SurfnetTransactionStatus, TransactionWithStatusMeta},
+};
+use chrono::Utc;
+use crossbeam_channel::{Receiver, Sender};
+use litesvm::{types::TransactionResult, LiteSVM};
+use solana_account::Account;
+use solana_client::{nonblocking::rpc_client::RpcClient, rpc_response::RpcPerfSample};
+use solana_clock::{Clock, Slot};
+use solana_epoch_info::EpochInfo;
+use solana_feature_set::{disable_new_loader_v3_deployments, FeatureSet};
+use solana_keypair::Keypair;
+use solana_message::{Message, VersionedMessage};
+use solana_pubkey::Pubkey;
+use solana_sdk::{
+    bpf_loader_upgradeable::get_program_data_address, system_instruction,
+    transaction::VersionedTransaction,
+};
+use solana_signature::Signature;
+use solana_signer::Signer;
+use solana_transaction_status::{
+    EncodedConfirmedTransactionWithStatusMeta, TransactionStatus, UiTransactionEncoding,
+};
+use std::collections::{HashMap, VecDeque};
+use surfpool_types::{
+    SimnetEvent, TransactionConfirmationStatus, TransactionMetadata, TransactionStatusEvent,
+};
+
+// #[cfg(clippy)]
+// const SUBGRAPH_PLUGIN_BYTES: &[u8] = &[0];
+
+// #[cfg(not(clippy))]
+// const SUBGRAPH_PLUGIN_BYTES: &[u8] =
+//     include_bytes!("../../../../target/release/libsurfpool_subgraph.dylib");
+
+pub type AccountFactory = Box<dyn Fn(&SurfnetSvm) -> Account + Send + Sync>;
+
+pub enum GetAccountStrategy {
+    LocalOrDefault(Option<AccountFactory>),
+    ConnectionOrDefault(Option<AccountFactory>),
+    LocalThenConnectionOrDefault(Option<AccountFactory>),
+}
+
+pub enum GeyserEvent {
+    NewTransaction(VersionedTransaction, TransactionMetadata, Slot),
+}
+
+/// `SurfnetSvm` provides a lightweight Solana Virtual Machine (SVM) for testing and simulation.
+///
+/// It supports a local in-memory blockchain state,
+/// remote RPC connections, transaction processing, and account management.
+///
+/// It also exposes channels to listen for simulation events (`SimnetEvent`) and Geyser plugin events (`GeyserEvent`).
+pub struct SurfnetSvm {
+    pub inner: LiteSVM,
+    pub transactions: HashMap<Signature, SurfnetTransactionStatus>,
+    pub perf_samples: VecDeque<RpcPerfSample>,
+    pub transactions_processed: u64,
+    pub connection: SurfnetDataConnection,
+    pub latest_epoch_info: EpochInfo,
+    pub simnet_events_tx: Sender<SimnetEvent>,
+    pub geyser_events_tx: Sender<GeyserEvent>,
+}
+
+pub enum SurfnetDataConnection {
+    Local,
```

Comment:
Once I started to grok the changes, I think I found this enum variant name to be misleading. Do you think `Uninitialized` would be more accurate?

Reply:
I'll go with `Offline` and added a few guard conditions in the code